RI Claim Embedded Tokens (#17)
ga4gh-duri/ga4gh-duri.github.io: researcher_ids/RI_Claims_V1.md

    following structure encodes a "ga4gh.ControlledAccessGrants" RI Claim:

    ```
    "ga4gh" : {
      "ControlledAccessGrants": [
        <RI Claim Embedded Token (see definition below)>,
        <RI Claim Embedded Token (if more than one)>,
        ...
      ]
    }
    ```

    RI Claims can be bundled together in a [Passport](#passport).

#### **Researcher Identity Claim Embedded Token ("RI Claim Embedded Token")**

-   An Assertion from a [Claim Source](#claim-source) that is bound to a
    researcher identity and signed by an
    [Embedded Claim Signatory](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-embedded-claim-signatory)
    as per the "iss" field.

-   Encoded as an
    [Embedded Token](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-embedded-token)
    that contains various properties
    ([RI Claim Object Fields](#ri-claim-object-fields)) that describe the assertion and
    limitations thereof.
    
-   RI Claim Embedded Tokens are encoded as strings and included as a list of
    strings returned within a [RI Claim](#researcher-identity-claim-ri-claim). See

>     the [RI Claim Embedded Token](#ri-claim-embedded-token) section of this

Reviewer: The name of the current section _Researcher Identity Claim Embedded Token ("RI Claim Embedded Token")_ is the same as the name of the referenced section _RI Claim Embedded Token_  which is a bit confusing.

Author: Renamed the section that describes the format to use "format" in its title.

Author: Renamed sections.

    specification for more details.

#### **Researcher Identity Claim Object ("RI Claim Object")**

> An OIDC claim "ga4gh_rio" that contains fields that provide GA4GH

Reviewer: The ga4gh_rio claim **is not**  an OIDC claim. It is a JWT claim as specified in the section 2 "Terminology" of RFC 7519 "JSON Web Token", see https://tools.ietf.org/html/rfc7519#section-2
OIDC claims as defined in https://openid.net/specs/openid-connect-core-1_0.html#Claims are defined as members of JSON object returned by /userinfo or in id_token.

Thus this section should be reworded so that "OIDC" should be replaced with "JWT", and it should mention the _JSON Web Token Claims Registry_ 
https://tools.ietf.org/html/rfc7519#section-10.1 instead of OIDC specification.

Author: Done

standard information about a claim that is not captured in the standard
set of claims published by underlying OIDC specification.

For field definitions, refer to [RI Claim Object Fields](#ri-claim-object-fields).

#### **RI Claim Object Identity**

> -   The {"iss", "sub"} pair of OIDC standard claims that are included

Reviewer: Again, "iss" and "sub" are not OIDC standard claims, they are JWT standard claims defined in the JWT specification https://tools.ietf.org/html/rfc7519#section-4.1.1

Author: Done

    within an [RI Claim Object](#researcher-identity-claim-object) that
    represents a given user (such as a user account) within the "iss"
    system.

#### **Claim Source**

-   The
    [AAI Claim Source](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-claim-source)
    as encoded by the [source field](#source).

> #### **Passport**

Reviewer 2: I know that a "Passport" is a somewhat conceptual idea, though its presence in the specifications always feels like it should be a specific data object with a specific data structure. I think from a data structure perspective we're referring to a bundle of claims as a token? Should we reconcile that language somewhat?

Author: Although probably not what you were expecting, there has been a large overhaul of definitions including this one, as well as an Overview section to try to tie it together better.

-   A logical collection of RI Claims that includes claims on the
    [Passport Access Token](#passport-access-token) as well as any

>     indirect RI Claims that such a JWT token makes available to the

Reviewer: A **JWT token** does not make anything available to the caller of /userinfo endpoint. Only an **access token** can do that. Thus the word "JWT" should be replaced with the word "access".

Author: Done

    caller of the corresponding /userinfo endpoint.

#### **Passport Access Token**

-   A
    [specialized JWT access token](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#ga4gh-jwt-format)
    that is signed by an
    [Identity Broker](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-identity-broker)
    as per the
    [GA4GH AAI specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md)
    for the purpose of encoding identity and evaluating authorization.
    

> -   A Passport Access Token does not contain any

Reviewer: This is contrary to the text of the previous section titled "Passport".

Author: Good catch, thanks. Fixed.

"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

### **Objects and Tokens**

#### Passport

-   A GA4GH-compatible access token, as per the [GA4GH AAI
    specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md),
    along with the [Passport JWT Claim](#passport-jwt-claim) that is
    returned from [Passport Broker](#passport-broker) endpoints using such an
    access token.

#### Passport JWT Claim

-   A [GA4GH
    Claim](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-ga4gh-claim)
    that is encoded as a list of [Passport Visas](#passport-visa)

>     provided by a common key value within the `ga4gh_passport_v1` JWT claim.

Reviewer: This line says "provided by a common key value within the `ga4gh_passport_v1` JWT claim", but the example immediately below has no keys under the claim, its value is a JSON list.

Author: Fixed.

    [AAI Specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md).

8.  <a name="requirement-8"></a> Passport Visas are designed for machine
    interpretation only to make an access decision and is a non-goal to support
    rich user interface requirements nor do these claims fully encode the audit
    trail.

9.  <a name="requirement-9"></a> A Passport Visa Object MAY
    contain a "[condition](#condition)" field that restricts the Passport Visa
    to only be valid when the condition is met.

    -   For example, an identity can have several affiliations and a
        "ControlledAccessGrants" in a Passport Visa MAY be coupled to one
        of them using the Condition field.

10. <a name="requirement-10"></a> Processing a Passport within a Passport
    Clearinghouse is to abide by the following:

    1.  A Passport Clearinghouse MUST ignore all Passport Visas is does not
        need to process a particular request.

    2.  Passport Clearinghouses MUST ignore Passport Visas unless:
    
        1.  The Passport Clearinghouse has a sufficient trust relationship
            with the [Passport Assertion Source](#passport-assertion-source);

>             or

Reviewer: The Clearing house should have a trust in both the Passport Assertion Source and the Passport Visa Signatory. Otherwise an evil Passport Visa Signatory may issue fake assertions pretending to be from a trusted Passport Assertion Source.

Author: Done

"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

### **Objects and Tokens**

#### Passport

> -   A GA4GH-compatible access token, as per the [GA4GH AAI

Reviewer 3: It would be elegant to focus on claim syntax and semantics in this spec and leave all protocol stuff ("access token") to the AAI spec.

Author: As discussed, looking to address this issue in a following PR.

    but Passport Visas do not generally indicate individual persons involved
    in making the assertion (i.e. who assigned/signed the claim) as detailed
    identity information is not needed to make an access decision.

5.  <a name="requirement-5"></a> Additional information about identity
    that is not needed to make an access decision SHOULD not be included in the
    Passport Visas. Identity description, encoding audit details, other data
    for non-access purposes are not the intent of these Passport Visas,
    and must be handled via other means beyond the scope of this specification
    should they be needed for entities and systems with sufficient authority to
    process such information.

6.  <a name="requirement-6"></a> All Passport Visas within the
    `ga4gh_passport_v1` scope eligible for release to the requestor MUST be
    consistently provided. Reasons for limiting exchange may include user
    approval, contractual limitations, regulatory restrictions, or filtering
    claims to only the subset needed for a particular purpose, etc.

7.  <a name="requirement-7"></a> The [Passport JWT Claim](#passport-jwt-claim)

>     is only included in the Passport if the `ga4gh_passport_v1` scope is

Reviewer 3: It would be elegant to make this spec focus on RI claims and leave all protocol stuff ("scope") to the AAI spec.

Author: Required to specify the name of the passport scope as the AAI spec leaves this as an abstraction to handle more than one spec using it.

    requested from the [Passport Broker](#passport-broker) and other conditions
    are met as outlined within this specification.

    If the Passport Broker is the [Passport Visa
    Signatory](#passport-visa-signatory), it MUST set the "iss" to itself and

>     sign such Embedded Tokens with its own private key as described in the

Reviewer 3: ```suggestion
    sign such Passport Visas with its own private key as described in the
```

Author: Done

    [AAI Specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md).

8.  <a name="requirement-8"></a> Passport Visas are designed for machine
    interpretation only to make an access decision and is a non-goal to support
    rich user interface requirements nor do these claims fully encode the audit
    trail.

9.  <a name="requirement-9"></a> A Passport Visa Object MAY
    contain a "[condition](#condition)" field that restricts the Passport Visa
    to only be valid when the condition is met.

    -   For example, an identity can have several affiliations and a
        "ControlledAccessGrants" in a Passport Visa MAY be coupled to one
        of them using the Condition field.

10. <a name="requirement-10"></a> Processing a Passport within a Passport
    Clearinghouse is to abide by the following:
    

>     1.  Claim Clearinghouses MUST reject all requests that provide Passports

Reviewer 3: ```suggestion
    1.  Passport Clearinghouses MUST reject all requests that provide Passports
```

Author: Done

Author: Done

        that fail the neccessary checks of the access token as described in
        the [GA4GH AAI
        Specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md).

    2.  A Passport Clearinghouse MUST ignore all Passport Visas is does not
        need to process a particular request.

    3.  Passport Clearinghouses MUST ignore Passports and Passport Visas
        unless:
    
        1.  The Passport Clearinghouse has a sufficient trust relationships
            with all of: the [Passport Broker](#passport-broker), [Passport
            Assertion Source](#passport-assertion-source),
            [Passport Visa Signatory](#passport-visa-signatory); or

        2.  The Passport Clearinghouse can rely on a trusted service that

>             provides trust sufficient trust of any of the Passport Broker,

Reviewer 3: ```suggestion
            provides sufficient trust of any of the Passport Broker,
```

Author: Done

    2.  The Identity Broker populates the "ga4gh_userinfo_claims" OIDC claim
        on the token as well as other claims as per the
        [GA4GH Access Token format](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#access-token-issued-by-broker),
        and MUST NOT include the "ga4gh" claims within the Passport Access
        Token. The RI Claims are only available at the /userinfo endpoint
        based on the "scope" claim as outlined elsewhere within this
        specification.

>     3.  The Identity Broker introduces new RI Claim Embedded Tokens from a

Reviewer: This sentence is confusing. A broker does not introduce embedded tokens by encoding claim objects, it simply includes embedded tokens received from Embedded Token Signatory without any modification into the RI Claims.

Author: Improved the description.

        [Claim Signatory](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-embedded-claim-signatory)
        by encoding RI Claim Objects within the "ga4gh" OIDC claim (available
        via /userinfo). If the Broker is the Claim Signatory, it MUST set
        the "iss" to itself and sign such Embedded Tokens with its own
        private key as described in the
        [AAI Specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md).

    4.  The Identity Broker propagates RI Claims that it received

>         from upstream Brokers by copying each RI Claim Object Embedded

Reviewer: ```suggestion
        from upstream Brokers by copying each RI Claim Embedded
```

Author: Done

        Token string into the coorsponding RI Claim within the outer "ga4gh"
        OIDC claim. These strings are provided unmodified, however they
        MAY be filtered as described elsewhere in this specification.

>     5.  The Identity Broker signs the JWT, making it a Passport Access

Reviewer: The section v. does not make sense at all. The broker returns RI Claims in the /userinfo response, which is not a JWT, and it cannot possibly be an access token.

Author: This is the signing of the Passport Access Token, which is the Broker's responsibility. Reordered and rephrased some of this section to simplify to some degree.

        Token.

>     6.  The Passport for the /userinfo request MUST contain "ga4gh" as a

Reviewer: The "Passport" as defined above is not an access token, thus "the passport for the /userinfo request" is a non-defined thing.
It should probably be "access token", but in fact there is no need to have the "scope" claim inside of the access token, it is not used by the Authorization Server itself.

Author: Rephrased

8.  <a name="requirement-8"></a> RI Claims are designed for machine
    interpretation only to make an access decision and is a non-goal to support
    rich user interface requirements nor do these claims fully encode the audit
    trail.

9.  <a name="requirement-9"></a> An RI Claim Embedded Token MAY
    contain a "[condition](#condition)" field that restricts the RI Claim
    Embedded Token to only be valid when the condition is met.

    -   For example, an identity can have several affiliations and a
        ControlledAccessGrants in a RI Claim Embedded Token MAY be
        coupled to one of them using the Condition field.

10. <a name="requirement-10"></a> Processing a Passport within a Claim
    Clearinghouse is to abide by the following:

    1.  A Claim Clearinghouse MUST ignore all RI Claim Embedded Tokens
        is does not need to process a particular request and MUST ignore all
        RI Claim Embedded Tokens unless it explicitly has a sufficient trust
        relationship with the "[source](#source)" of the RI Claim Embedded Token.

    2.  Claim Clearinghouses SHOULD ignore claims that aren’t needed for their
        purposes.
        

>     3.  When combining RI Claim Objects from multiple RI Claim Accounts for

Reviewer: ```suggestion
    3.  When combining RI Claim Embedded Tokens from multiple RI Claim Sources for
```

Author: Rephrased

-   May provide a rich set of internationalization/localization features for
    clients to consume.

## RI Claim Embedded Token

Each [RI claim](#researcher-identity-claim-ri-claim) name maps to an array of
([RI Claim Embedded Tokens](#researcher-identity-claim-embedded-token-ri-claim-embedded-token))
which are encoded as
[Embedded Tokens](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-embedded-token)
within a "ga4gh" root OIDC claim object.

Non-normative example of a set of RI Claim Embedded Tokens, encoded as Embedded
Token strings, with a RI Claim named "ResearcherStatus":

```
{
  "ga4gh": {
    "ResearcherStatus": [
      "eyJhbGciOiJI...aaa",
      "eyJhbGciOiJI...bbb"
    ]
  }
}
```

An [example](#example-ri-claims) of encoding multiple RI Claims in a more
reader-friendly form is also available.

### RI Claim Embedded Token Requirements

-   RI Claim Embedded Tokens MUST conform to the
    [GA4GH AAI Specification Embedded Token format](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#embedded-token-issued-by-broker)

>     as a base64 string.

Reviewer: Is is not a base64 string, it is a "JWS Compact Serialization" string as defined at https://tools.ietf.org/html/rfc7515#section-7.1 i.e. it is 3 base64 strings joined by dots.

Author: Done

-   Claim Signatories, Identity Brokers, Claim Clearinghouses, and their
    Clients MUST conform to the
    [GA4GH AAI Specification](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md)
    requirements for Embedded Tokens in their use of RI Claim Embedded Tokens.

-   Validation, as outlined elsewhere in this specification and the
    GA4GH AAI Specification, MUST be performed before RI Claim
    Embedded Tokens are used for identity or authorization.

-   The "jti" is REQUIRED.

-   If the OIDC Access Token format is used (i.e. the "scope" claim
    contains the "openid" scope) as per the GA4GH AAI Specification's
    Embedded Token encoding options, then the /userinfo endpoint is unused
    but the /introspect endpoint MUST be available for
    [claim polling](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#claim-polling).

> 

Reviewer: It is not a requirement for an RI Claim Embedded Token to contain an "RI Claim Object" in the ga4gh_rio JWT claim?

Author: Removed this stuff as it is now covered better by the AAI spec.

### RI Claim Embedded Token Format

```
{
  "typ": "JWT",
  "alg": "RS256",
  ["jku": "\<JKS URL\>",]
  "kid": "\<Key Identifier\>"
}.
{
  "iss": "\<Issuer URL\>",
  "sub": "\<Subject\>",
  ["scope": "openid ...",]
  "jti": "\<JTI Token Identifier\>",
  "iat": \<Seconds since epoch\>,
  "exp": \<Seconds since epoch\>,
  "ga4gh_rio": {
    "value": "\<Value String\>",
    "source": "\<Source URL\>",
    ["condition": {...},]
    ["by": "\<By identifier\>"]
  }
}.\<Signature\>
```

Where fields within the `ga4gh_rio` object are as described in the
[RI Claim Object Fields](#ri-claim-object-fields) section of this
specification except for "iat" and "exp" which are direct claims on
the outer JWT itself.

One of `scope` or `jku` MUST be present as described in

> [RI Claim Embedded Token Requirements](#ri-claim-embedded-token-requirements).

Reviewer: The referenced section does not contain the mentioned requirement.

Author: Fixed.

    clients to consume.

## Passport JWT Claim Format

The [Passport JWT Claim](#passport-jwt-claim) name maps to an array of
[Passport Visas](#passport-visa) which are encoded as
[Embedded Tokens](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#term-embedded-token)
within a Passport JWT Claim.

Non-normative example of a set of Passport Visas, encoded as Embedded
Token JWS Compact Serialization strings:

```
{
  "ga4gh_passport_v1": [
    "<eyJhbGciOiJI...aaa>",
    "<eyJhbGciOiJI...bbb>"
  ]
}
```

For a more reader-friendly example, see the [Example Passport JWT
Claim](#example-passport-jwt-claim) section of the specification.

### Passport Visa Requirements

-   Passport Visas MUST conform to one of the
    [GA4GH AAI Specification Embedded Token formats](https://github.com/ga4gh/data-security/blob/master/AAI/AAIConnectProfile.md#embedded-token-issued-by-embedded-token-signatories)
    as JWS Compact Serialization string as defined by [RFC7515 section
    7.1](https://tools.ietf.org/html/rfc7515#section-7.1).

-   Embedded Token Signatories, Passport Brokers, Passport Clearinghouses, and

>     their Clients MUST conform to the

Reviewer 3: "and their clients" Where does the scope of this spec end? I don't think we should impose requirements to actors who are not covered by this document

Author: Limited the scope of client conformance to just the part that the AAI spec calls out.

    [Passport Assertion Source](#passport-assertion-source) requires such a
    claim to be no longer valid. A Passport Assertion Source MAY choose
    to make a claim very long lived. However, a [Passport Visa
    Signatory](#passport-visa-signatory) MAY choose an earlier timestamp
    in order to limit the claim’s duration within downstream Passport
    Clearinghouses.

-   Access is NOT necessarily removed by the `exp` timestamp. Instead,
    this timestamp may be viewed as a cut-off after which no new access
    will be granted and action to remove any existing access may
    commence anytime shortly after this cut-off period.

-   Its use by a Passport Clearinghouse is described in the
    [Passport Visa Expiry](#passport-visa-expiry) section and
    [Token Revocation](#token-revocation) section.

## Passport Visa Fields

Although standard fields within a [Passport Visa Object](#passport-visa-object)
are defined in this section, other fields MAY exist within the object
and should be ignored by any Passport Clearinghouse that is not familiar
with the use of such fields. Field names are reserved for definition by

> the GA4GH DURI committee.

Reviewer 3: Can we assume DURI exists forever (or at least as long as this spec?) Would it be more future-proof to refer to GA4GH only?

Author: Opened an issue to track this

#### "**type**"

-   REQUIRED. A [Passport Visa Type](#passport-visa-type) that is
    either a [GA4GH Standard Passport Visa
    Definition](#ga4gh-standard-passport-visa-definition) name, or a
    [Custom Passport Visa Type](#custom-passport-visa-types) name.

#### "**asserted**"

-   REQUIRED. Seconds since unix epoch that represents when the [Passport
    Assertion Source](#passport-assertion-source) made the claim.

-   Its use by a Passport Clearinghouse is described in the
    [Passport Visa Expiry](#passport-visa-expiry) section.

-   If a [Passport Assertion Repository](#passport-assertion-repository)

>     does not include enough information to construct an `iat` timestamp, a

Reviewer 3: Can you cross-check the use of "asserted" and "iat". Is the difference between the two described somewhere?

Author: Fixed text in the spec. There are more details in the landing page.
